modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManagerImpl.java

```diff
@@ -3276,6 +3543,51 @@ else if (flags == TOMBSTONES)
             return cursor;
         }
 
+        /** {@inheritDoc} */
+        @Override public GridCursor<? extends CacheDataRow> reconCursor(int cacheId,
+            KeyCacheObject lower,
+            KeyCacheObject upper,
+            CacheDataRowAdapter.RowData x,
+            MvccSnapshot snapshot,
+            int flags
+        ) throws IgniteCheckedException {
+            SearchRow lowerRow;
+            SearchRow upperRow;
+
+            if (grp.sharedGroup()) {
+                assert cacheId != CU.UNDEFINED_CACHE_ID;
+
+                lowerRow = lower != null ? new SearchRow(cacheId, lower) : new SearchRow(cacheId);
+                upperRow = upper != null ? new SearchRow(cacheId, upper) : new SearchRow(cacheId);
+            }
+            else {
+                lowerRow = lower != null ? new SearchRow(CU.UNDEFINED_CACHE_ID, lower) : null;
+                upperRow = upper != null ? new SearchRow(CU.UNDEFINED_CACHE_ID, upper) : null;
+
+                cacheId = CU.UNDEFINED_CACHE_ID;
+            }
+
+            GridCursor<? extends CacheDataRow> cursor;
+
+            if (snapshot != null) {
+                assert grp.mvccEnabled();
+
+                GridCacheContext cctx = grp.sharedGroup() ? grp.shared().cacheContext(cacheId) : grp.singleCacheContext();
+
+                cursor = dataTree.find(lowerRow, upperRow, new MvccFirstVisibleRowTreeClosure(cctx, snapshot), x);
```

Comment:
Despite the fact that I'd rather remove reconCursor() (see comment above), why we use find in case of snapshot!=null instead of reconFind()?

Reply:
The reconciliation cursor is always created with 'snapshot == null`. I added `assert flags != RECONCILIATION`.